pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1686 @@
+			ensure!(
+				pallet_uniques::Pallet::<T>::owned_in_class(&spirit_collection_id, &sender).count() >
+					0,
+				Error::<T>::MustOwnSpiritToPurchase
+			);
+
+			// Get preorder_id for new preorder
+			let preorder_id =
+				<PreorderIndex<T>>::try_mutate(|n| -> Result<PreorderId, DispatchError> {
+					let id = *n;
+					ensure!(id != PreorderId::max_value(), Error::<T>::NoAvailablePreorderId);
+					*n += 1;
+					Ok(id)
+				})?;
+			let overlord = Self::get_overlord_account()?;
+			// Verify metadata
+			let mint_metadata = Self::verify_nft_metadata(&overlord, metadata)?;
+			let preorder = PreorderInfo {
+				owner: sender.clone(),
+				race: race.clone(),
+				career: career.clone(),
+				metadata: mint_metadata,
+				preorder_status: PreorderStatus::Pending,
+			};
+			// Reserve currency for the preorder at the Hero origin_of_shell price
+			<T as pallet::Config>::Currency::reserve(&sender, T::HeroOriginOfShellPrice::get())?;
+
+			Preorders::<T>::insert(preorder_id, preorder);
+
+			Self::deposit_event(Event::OriginOfShellPreordered { owner: sender, preorder_id });
+
+			Ok(())
+		}
+
+		/// This is an admin only function that will be used to set a preorder's status from Pending
+		/// to either Chosen or NotChosen
+		///
+		/// Parameters:
+		/// `origin`: Expected to come from Overlord admin account
+		/// `preorder_id`: Preorder id to change preorder status
+		/// `status`: Either Chosen or NotChosen
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn set_preorder_status(
+			origin: OriginFor<T>,
+			preorder_id: PreorderId,
+			status: PreorderStatus,
+		) -> DispatchResult {
+			// Ensure Overlord account makes call
+			let sender = ensure_signed(origin)?;
+			Self::ensure_overlord(sender)?;
+			// Ensure value exists
+			ensure!(Preorders::<T>::contains_key(preorder_id), Error::<T>::NoAvailablePreorderId);
+			// Change status of preorder and add to PreorderResult
+			let mut preorder_info =
+				Preorders::<T>::take(preorder_id).ok_or(Error::<T>::NoAvailablePreorderId)?;
+			preorder_info.preorder_status = status.clone();
+			PreorderResults::<T>::insert(&preorder_info.owner, preorder_id, &preorder_info);
+
+			Self::deposit_event(Event::PreorderResultChanged { preorder_id, status });
+
+			Ok(())
+		}
+
+		/// Claim function for bulk minting chosen preorders for the sender
+		///
+		/// Parameters:
+		/// `origin`: Sender
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn claim_chosen_preorders(origin: OriginFor<T>) -> DispatchResult {
+			ensure!(
+				!CanPreorderOriginOfShells::<T>::get(),
+				Error::<T>::PreorderOriginOfShellNotAvailable
+			);
+			let sender = ensure_signed(origin)?;
+			let overlord = Self::get_overlord_account()?;
+
+			// Has Spirit Collection been set
+			let spirit_collection_id = Self::get_spirit_collection_id()?;
+			// Must have a spirit to purchase a Origin of Shell
```

Comment:
I think this one is the overlord?

Reply:
For this the `overlord` account is used to mint the Origin of Shell Preorder. The `sender` will be the user claiming their `Chosen` Preorders for minting

Reply:
`OverlordOrigin` Trait has been updated and set in runtime to `EnsureRootOrHalfCouncil`